src/BilinearInterpolation.jl

```diff
@@ -0,0 +1,129 @@
+module BilinearInterpolation
+using DocStringExtensions
+import ..Interpolation1D: get_stencil, Linear, Flat
+
+"""
+    Bilinear{V,I}
+
+This struct stores the source and target grids for bilinear interpolation on a rectangular grid.
+Indexes specifying the location of the target points in the source grid are also stored for 
+enabling efficient implementation of bilinear interpolation.
+
+# Fields
+$(DocStringExtensions.FIELDS)
+"""
+struct Bilinear{V,I}
+    "Source grid in `x` direction `(nsourcex)`"
+    sourcex::V
+    "Source grid in `y` direction `(nsourcey)`"
+    sourcey::V
+    "Target grid in `x` direction `(ntargetx)`"
+    targetx::V
+    "Target grid in `y` direction `(ntargety)`"
+    targety::V
+    "left index of location of a target `x` grid point in source `x` grid"
+    startx::I
+    "left index of location of a target `y` grid point in source `y` grid"
+    starty::I
+end
+
+get_dims(b::Bilinear) =
+    (length(b.sourcex), length(b.sourcey), length(b.targetx), length(b.targety))
+
+function Bilinear(sourcex, sourcey, targetx, targety)
+    nsourcex, nsourcey = length(sourcex), length(sourcey)
+    ntargetx, ntargety = length(targetx), length(targety)
+    startx = similar(targetx, Int)
+    starty = similar(targety, Int)
+    set_source_range!(startx, sourcex, targetx)
+    set_source_range!(starty, sourcey, targety)
+    return Bilinear(sourcex, sourcey, targetx, targety, startx, starty)
+end
+
+function set_source_range!(
+    start::AbstractVector{I},
+    source::AbstractVector{FT},
+    target::AbstractVector{FT},
+) where {I,FT}
+    order = Linear()
+    first = 1
+    @inbounds begin
+        for (ixt, x) in enumerate(target)
+            st = get_stencil_bilinear1d(order, source, x, first = first)
+            first = st
+            start[ixt] = st
+        end
+    end
+    return nothing
+end
+
+"""
+    interpolatebilinear!(
+        ftarget::AbstractArray{FT,N},
+        bilinear::B,
+        fsource::AbstractArray{FT,N},
+    ) where {FT,N,B}
+
+Interpolate `fsource`, defined on source grid, onto the target grid.
+The horizontal source and targets are define in `bilinear`.
+Here `fsource` is an N-dimensional array where the last two dimensions are assumed to be horizontal dimensions.
+
+For example, `fsource` can be of size `[n1, n2..., nx, ny]`, where `nx` and `ny` are the horizontal dimensions.
+Single horizontal level is also supported.
+The number of horizontal levels should be same for both source and target arrays.
+"""
+function interpolatebilinear!(
+    ftarget::AbstractArray{FT,N},
+    bilinear::B,
+    fsource::AbstractArray{FT,N},
```

Comment:
Shouldn't we allow the arrays for `ftarget` and `fsource` to have different lengths, rather than both being `N`? E.g. if interpolating from low resolution to high resolution

Reply:
It does. Here, `N` is just the dimension of the array!
It expects the source functions and target functions to be of sizes `(n1, n2, ..., nh1_source, nh2_source)` and `(n1, n2, ...., nh2_target, nh2_target)` respectively.

Reply:
Please take a look at the test here: `test/bilinearinterpolation.jl`